core/src/main/java/com/linecorp/armeria/client/limit/AbstractConcurrencyLimitingClient.java

```diff
@@ -118,73 +101,33 @@ public final int numActiveRequests() {
 
     @Override
     public final O execute(ClientRequestContext ctx, I req) throws Exception {
-        return maxConcurrency == 0 ? unlimitedExecute(ctx, req)
-                                   : limitedExecute(ctx, req);
-    }
-
-    private O limitedExecute(ClientRequestContext ctx, I req) throws Exception {
         final CompletableFuture<O> resFuture = new CompletableFuture<>();
         final O deferred = newDeferredResponse(ctx, resFuture);
-        final PendingTask currentTask = new PendingTask(ctx, req, resFuture);
-
-        pendingRequests.add(currentTask);
-        drain();
-
-        if (!currentTask.isRun() && timeoutMillis != 0) {
-            // Current request was not delegated. Schedule a timeout.
-            final ScheduledFuture<?> timeoutFuture = ctx.eventLoop().withoutContext().schedule(
-                    () -> resFuture.completeExceptionally(
-                            UnprocessedRequestException.of(ConcurrencyLimitTimeoutException.get())),
-                    timeoutMillis, TimeUnit.MILLISECONDS);
-            currentTask.set(timeoutFuture);
-        }
-
+        concurrencyLimit.acquire(ctx)
+                        .whenCompleteAsync((permit, throwable) -> {
+                            if (throwable != null) {
+                                resFuture.completeExceptionally(throwable);
+                                return;
+                            }
+                            numActiveRequests.incrementAndGet();
+                            try (SafeCloseable ignored = ctx.replace()) {
+                                try {
+                                    final O actualRes = unwrap().execute(ctx, req);
+                                    actualRes.whenComplete().whenCompleteAsync((unused, cause) -> {
```

Comment:
How about using `handle` here? Looks like we don't need `handleAsync` to me. Please feel free to correct me though.

Reply:
Here was to pin it to the same `eventLoop`

Reply:
Do we need to? Aren't `permit.close()` and `numActiveRequests.decrementAndGet()` thread-safe?

Reply:
Yes, they are, I wasn't sure if it would context switch.